kong/plugins/apifortress/fortress_serializer.lua

```diff
@@ -0,0 +1,117 @@
+-- The serializer. Builds a data structure that is compatible with
+-- the API Fortress API expectations
+
+local stringy = require "stringy"
+
+local _M = {}
+
+-- pre-indexing global functions
+local str_startswith = stringy.startswith
+local str_endswith = stringy.endswith
+local str_sub = string.sub
+local str_len = string.len
+local table_insert = table.insert
+local encode_base64 = ngx.encode_base64
+
+function _M.serialize(ngx)
+
+  -- pre-indexing to reduce the number of lookups later on
+  local ngx_req = ngx.req
+  local ngx_resp = ngx.resp
+  local ngx_var = ngx.var
+  local ngx_ctx = ngx.ctx
```

Comment:
Then there's something I'm not understanding.
At the top level, I indexed the variables / functions with no state, as in: `local encode_base64 = ngx.encode_base64` .
I understand this is safe as demonstrated in alf.lua, as you pointed out.

In the `serialize` function, I'm preindexing locally all the things that are going to be used in the same function as in: `local ngx_req = ngx.req` to drastically reduce the lookups.
Now, in Lua documentation I read:
> local variables have their scope limited to the block where they are declared. A block is the body of a control structure, the body of a function, or a chunk (the file or string with the code where the variable is declared). 

So I'm guessing (but I might be wrong of course) that in my code `local ngx_ctx = ngx.ctx` as part of the `serialize` function will create an `ngx_ctx` variable with a scope that is bound to the function itself.